README.md
Python 3 modules for accessing .dbf files
------------------------------------------

### Features

* Support dBase, FoxPro...
* Support memo
* Code page

### TODO

* Write test

### Install

1. Download [master.zip](https://github.com/SteelyWing/dbfpy/archive/master.zip) and extract.
2. Run `python setup.py`

### Demo

See [demo.py](https://github.com/SteelyWing/dbfpy/blob/master/examples/demo.py)

---

### NOT FULLY TEST YET

### USE AT YOUR OWN RISK

### NO WARRANTIES WHATSOEVER

---

Jeff Kunce <kuncej@mail.conservation.state.mo.us>

http://starship.python.net/crew/jjkunce/

Hans Fiby <hans@fiby.at>

http://www.fiby.at


CHANGES
Version 2.2.5 (16-sep-2010)
---------------------------

Fix Y2K issue with Last Update field in the header (sf bug 3065838).

Version 2.2.4 (26-may-2009)
---------------------------

Ignore leading and trailing zero bytes in numeric fields.

Version 2.2.3 (17-feb-2009)
---------------------------

Support for writing empty date values.

Version 2.2.2 (16-sep-2008)
---------------------------

Numeric decoder returns float when the value contains decimal point
without regard to the number of decimal places declared in the header.

Version 2.2.1 (16-mar-2008)
---------------------------

Fix: raise ValueError if a name passed to field constructor
is longer than 10 characters.

Version 2.2.0 (11-feb-2007)
---------------------------

Features:
  - date fields are allowed to contain values stored with leading spaces
    instead of leading zeroes.
  - dbf.header returns field definition objects if accessed as a list
    or as a dictionary.
  - added raw data access methods: DbfRecord.rawFromStream(),
    DbfFieldDef.rawFromRecord().
  - added conversion error handling: if ignoreErrors=True is passed
    with Dbf constructor arguments, then failing field value conversions
    will not raise errors but instead will return special object
    INVALID_VALUE which is equal to None, empty string and zero.
    .ignoreErrors property of the Dbf instances may be toggled also
    after instance initialization.

Version 2.1.0 (01-dec-2006)
---------------------------

Features:
  - support field types 'F' (float), 'I' (integer) and 'Y' (currency)

Fix processing of empty Timestamp values

Version 2.0.3 (30-oct-2006)
---------------------------

Fixes:
  - compatibility fix for Python versions prior to 2.4 (sf bug 1574526)
  - wrong record length when reading from file object (sf bug 1586619)

Version 2.0.2 (08-jul-2006)
---------------------------

Fix dbfnew (legacy API for DBF creation)

Version 2.0.1 (10-mar-2006)
---------------------------

Fixes in Numeric field processing:
  - decoding inserted decimal point where it shouldn't;
  - encoded value never exceeds field length.
    If encoded value is too large for the field, decimal digits get
    removed (rounding the value down).  When integer value is larger
    than the field length, ValueError is raised.

Fix: Date and Logical fields were not able to decode empty values.

Errors raised from field processing include field name.

This file is included in the source distribution.

Version 2.0.0 (20-dec-2005)
---------------------------

Changed style of doc-strings; added more docs.

``mx.DateTime`` isn't required anymore - internal representation uses
``datetime.date`` and ``datetime.datetime`` classes from the batteries.

'D' and 'T' became "smarter":
  - date fields could accept:
    - date/time tuples [or tuples with at least
      first three components (year, day, month)];
    - strings in format 'yyyymmdd' or 'yymmdd';
    - ``datetime.date`` and ``datetime.datetime`` instances;
    - ``None`` instance - use current date value;
    - numbers (``int``, ``long``, ``float``)
      which are treated as seconds from epoch;
    - objects having 'ticks' method (``mx.DateTime`` for example);
  - datetime fields can accept everything date fields do
    except string objects (which is a TODO stuff).

Changed identifiers and args of some methods. For example old code

         dbf = Dbf()
         dbf.openFile("county.dbf", readOnly=1)

now should be written like

         dbf = Dbf("county.dbf", True)

All classes now are new-style classes (inherited from the ``object``).

Got rid of ``binnum`` and ``strutil`` modules,
using ``struct`` and string methods instead.

Group classes in different modules instead of having them all in
one, as it was earlier (see dbf.py from the previous release(s)).

Use fields' registry to simplify registering of the new data types.


Previous changes:
-----------------

took over development 2000-10-06 Hans Fiby

changes: 2001-02-07 Hans Fiby
 * read long integers as long
 * add this CHANGES File

changed: 2000-10-06 Hans Fiby
 * added dbfnew.py
 * fixed decimal points


dbfpy/dbf.py
#! /usr/bin/env python
"""DBF accessing helpers.

FIXME: more documentation needed

Examples:

    Create new table, setup structure, add records:

        dbf = Dbf(filename, new=True)
        dbf.addField(
            ("NAME", "C", 15),
            ("SURNAME", "C", 25),
            ("INITIALS", "C", 10),
            ("BIRTHDATE", "D"),
        )
        for (n, s, i, b) in (
            ("John", "Miller", "YC", (1980, 10, 11)),
            ("Andy", "Larkin", "", (1980, 4, 11)),
        ):
            rec = dbf.newRecord()
            rec["NAME"] = n
            rec["SURNAME"] = s
            rec["INITIALS"] = i
            rec["BIRTHDATE"] = b
            rec.store()
        dbf.close()

    Open existed dbf, read some data:

        dbf = Dbf(filename, True)
        for rec in dbf:
            for fldName in dbf.fieldNames:
                print '%s:\t %s (%s)' % (fldName, rec[fldName],
                    type(rec[fldName]))
            print
        dbf.close()

"""

__version__ = "$Revision: 1.9 $"[11:-2]
__date__ = "$Date: 2012/12/17 19:16:57 $"[7:-2]
__author__ = "Jeff Kunce <kuncej@mail.conservation.state.mo.us>"

__all__ = ["Dbf"]

from .header import DbfHeader
from . import memo
from .record import DbfRecord
from . import utils


class Dbf(object):
    """DBF accessor.

    FIXME:
        docs and examples needed (dont' forget to tell
        about problems adding new fields on the fly)

    """

    __slots__ = ("name", "header", "stream", "memo", "_ignore_errors")

    INVALID_VALUE = utils.INVALID_VALUE

    ## initialization and creation helpers

    def __init__(self, file, read_only=False, new=False, ignore_errors=False,
                 memo_file=None):
        """Initialize instance.

        Arguments:
            file:
                Filename or file-like object.
            read_only:
                if ``f`` argument is a string file will
                be opend in read-only mode; in other cases
                this argument is ignored. This argument is ignored
                even if ``new`` argument is True.
            new:
                True if new data table must be created. Assume
                data table exists if this argument is False.
            ignore_errors:
                if set, failing field value conversion will return
                ``INVALID_VALUE`` instead of raising conversion error.
            memo_file:
                optional path to the FPT (memo fields) file.
                Default is generated from the DBF file name.

        """
        if isinstance(file, str):
            # a filename
            self.name = file
            if new:
                # new table (table file must be
                # created or opened and truncated)
                self.stream = open(file, "w+b")
            else:
                # table file must exist
                self.stream = open(file, ("r+b", "rb")[bool(read_only)])
        else:
            # a stream
            self.name = getattr(file, "name", "")
            self.stream = file

        if new:
            self.header = DbfHeader()
        else:
            self.header = DbfHeader.parse(self.stream)

        # for IDE inspection
        self._ignore_errors = None

        self.ignore_errors = ignore_errors
        if memo_file:
            self.memo = memo.MemoFile(memo_file, readOnly=read_only, new=new)
        elif self.header.has_memo:
            self.memo = memo.MemoFile(memo.MemoFile.memo_file_name(self.name),
                                      readOnly=read_only, new=new)
        else:
            self.memo = None
        self.header.set_memo_file(self.memo)

    ## properties

    @property
    def closed(self):
        return self.stream.closed

    @property
    def record_count(self):
        return self.header.record_count

    @property
    def field_names(self):
        return [field.name for field in self.header.fields]

    @property
    def fields(self):
        return self.header.fields

    @property
    def ignore_errors(self):
        """Error processing mode for DBF field value conversion

        if set, failing field value conversion will return
        ``INVALID_VALUE`` instead of raising conversion error.

        """
        return self._ignore_errors

    @ignore_errors.setter
    def ignore_errors(self, value):
        """Update `ignore_errors` flag on the header object and self"""
        self.header.ignore_errors = self._ignore_errors = bool(value)

    ## interface methods

    def close(self):
        """Close the stream, write the end of record 0x1A and truncate"""
        self.flush()

        if self.stream.writable():
            # write SUB (ASCII 26) after last record
            self.stream.seek(
                self.header.header_length +
                self.header.record_count * self.header.record_length
            )
            self.stream.write(b"\x1A")
            self.stream.truncate()

        self.stream.close()

    def flush(self):
        """Flush data to the associated stream."""
        self.header.flush(self.stream)
        self.stream.flush()
        # flush if memo is not None
        if hasattr(self.memo, 'flush'):
            self.memo.flush()

    def new_record(self):
        """Return new record, which belong to this table."""
        return DbfRecord(self.header)

    def write_record(self, record):
        """Write data to the dbf stream.

        If ``record.index`` is None, this record will be appended to the
        records of the DBF this records belongs to; or replaced otherwise.
        """
        if not self.stream.writable():
            return

        if record.index is None:
            # we must increase record count before set index,
            # because set index will raise error if out of range
            self.header.record_count += 1
            record.index = self.header.record_count - 1

        self.stream.seek(record.position)
        self.stream.write(record.to_bytes())

    def append(self, record):
        """Append ``record`` to the database."""
        record.index = None
        self.write_record(record)

    def add_field(self, *defs):
        """Add field definitions.

        For more information see `header.DbfHeader.add_field`.
        """
        self.header.add_field(*defs)
        if self.header.has_memo:
            if not self.memo:
                self.memo = memo.MemoFile(
                    memo.MemoFile.memo_file_name(self.name), new=True)
            self.header.set_memo_file(self.memo)

    ## 'magic' methods (representation and sequence interface)

    def __str__(self):
        return "Dbf stream '%s'\n" % self.stream + str(self.header)

    def __len__(self):
        """Return number of records."""
        return self.record_count

    def __getitem__(self, index):
        """Return `DbfRecord` instance."""
        if isinstance(index, slice):
            return [self[i] for i in range(self.record_count)[index]]

        record = DbfRecord(
            self.header, index=index
        )
        record.read(self.stream)
        return record

    def __setitem__(self, index, record):
        """Write `DbfRecord` instance to the stream."""
        record.index = index
        self.write_record(record)

    #def __del__(self):
        #    """Flush stream upon deletion of the object."""
        #    self.flush()

# vim: set et sw=4 sts=4 :
